fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/impl/content/registry/Taggable2ObjectMapRegistryImpl.java

```diff
@@ -0,0 +1,143 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.content.registry;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import net.minecraft.resource.ResourceManager;
+import net.minecraft.resource.ResourceType;
+import net.minecraft.tag.Tag;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.content.registry.v1.util.Taggable2ObjectMap;
+import net.fabricmc.fabric.api.resource.ResourceManagerHelper;
+import net.fabricmc.fabric.api.resource.ResourceReloadListenerKeys;
+import net.fabricmc.fabric.api.resource.SimpleSynchronousResourceReloadListener;
+
+public class Taggable2ObjectMapRegistryImpl<K, V> implements Taggable2ObjectMap<K, V>, SimpleSynchronousResourceReloadListener {
+	private final Identifier reloadIdentifier;
+	private static final Collection<Identifier> RELOAD_DEPS = Collections.singletonList(ResourceReloadListenerKeys.TAGS);
+
+	private boolean tagsPresent = false;
+	private final List<Runnable> loader = new LinkedList<>(); // This is to preserve the order of how things are added and removed. This would be a lot simpler if we did not have to deal with Tags
+	private final Map<K, V> restorer = new HashMap<>();
+	private final BiConsumer<K, V> putter;
+	private final Consumer<K> remover;
+	private final Function<K, V> getter;
+
+	Taggable2ObjectMapRegistryImpl(String name, BiConsumer<K, V> putter, Consumer<K> remover, Function<K, V> getter) {
+		reloadIdentifier = new Identifier("fabric:private/" + name);
+		ResourceManagerHelper.get(ResourceType.SERVER_DATA).registerReloadListener(this);
+		this.putter = putter;
+		this.remover = remover;
+		this.getter = getter;
+	}
+
+	@Override
+	public void apply(ResourceManager manager) {
+		reload();
+		tagsPresent = true;
+	}
+
+	private void reload() {
+		restorer.forEach(putter);
+		loader.forEach(Runnable::run);
+	}
+
+	@Override
+	public V get(K key) {
+		return getter.apply(key);
+	}
+
+	@Override
+	public void add(K key, V value) {
+		loader.add(
+				() -> {
+					restorer.put(key, get(key));
+					putter.accept(key, value);
+				}
+		);
+
+		if (tagsPresent) {
+			reload();
+		}
+	}
+
+	@Override
+	public void add(Tag<K> tag, V value) {
+		loader.add(
+				() -> {
+					for (K key : tag.values()) {
+						restorer.put(key, get(key));
+						putter.accept(key, value);
+					}
+				}
+		);
+
+		if (tagsPresent) {
+			reload();
+		}
+	}
+
+	@Override
+	public void remove(K key) {
+		loader.add(
+				() -> {
+					restorer.put(key, get(key));
+					remover.accept(key);
+				}
+		);
+
+		if (tagsPresent) {
+			reload();
+		}
+	}
+
+	@Override
+	public void remove(Tag<K> tag) {
```

Comment:
Quick question: what happens if a tag is removed but a specific in the tag is explicitly added?

Reply:
answered on discord, short answer is the order is preserved.